### PR TITLE
[cni-cilium] migrate cilium-agent images to distroless

### PR DESCRIPTION
**Pull request by AndreyPavlovFlant**
## Description

`cilium-agent` is based on a distroless images.

## Why do we need it, and what problem does it solve?

We'd like to base our images on a distroless image.

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: cni-cilium
type: chore
summary: cilium-agent is now based on a distroless images.
impact: cilium-agent pods will restart.
impact_level: default
```

### CODE REVIEW

File: modules/021-cni-cilium/images/builder/werf.inc.yaml [hunk 1]
```diff
@@ -1,78 +1,22 @@
...
-  - dpkg -i /iptables/*.deb
-  - rm -rf /iptables
-  - ./iptables-wrapper-installer.sh --no-sanity-check
+# #####################################################################
+# Build artifact (based on runtime-artifact)(based on Ubuntu)
+# Based on https://github.com/cilium/cilium/blob/v1.14.5/images/builder/Dockerfile
 ---
-image: {{ $.ModuleName }}/{{ $.ImageName }}-artifact
-fromImage: {{ $.ModuleName }}/{{ $.ImageName }}-runtime-artifact
+image: {{ $.ModuleName }}/builder-artifact
```
Comment (apolovov) [inline modules/021-cni-cilium/images/builder/werf.inc.yaml:12]: Why not artifact?

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/static'...

File: modules/021-cni-cilium/images/base-cilium-dev/binary_replace.sh [hunk 2]
```diff
@@ -0,0 +1,120 @@
+#!/bin/bash
```
Comment (apolovov) [inline modules/021-cni-cilium/images/base-cilium-dev/binary_replace.sh:1]: consider using binary_replace.sh from agent.

File: modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    libc6-dev libc6-dev-amd64-cross libelf-dev libmnl-dev \
+    bison flex \
+    build-essential libtool ninja-build pkg-config git-buildpackage \
+    python2 python3 python3-pip virtualenv \
+    libnetfilter-conntrack-dev libnetfilter-conntrack3 libnfnetlink-dev
+  - apt-get satisfy -y --no-install-recommends "debhelper-compat (= 13), libnftnl-dev (>= 1.1.6)"
+  - wget -qO- https://apt.llvm.org/llvm-snapshot.gpg.key | tee /etc/apt/trusted.gpg.d/apt.llvm.org.asc
+  - apt-add-repository -y "deb http://apt.llvm.org/jammy/ llvm-toolchain-jammy-15 main"
+  # for debug
+  - apt-get install -y --no-install-recommends strace
```
Comment (apolovov) [inline modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml:64]: Feel free to add more debug tools.

File: modules/021-cni-cilium/images/README.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  - based on `BASE_CILIUM_DEV` image
+  - includes:
+    - src of iptables deb-package *(loaded from fox)*
+    - deb-packages of iptables 1.8.8-1 *(builded from src)*
+- `+` `cilium-envoy`
+  - based on `BASE_CILIUM_DEV` image
+  - includes:
+    - src of cilium/proxy *(loaded from fox)*
+    - src of envoyproxy/envoy *(loaded from fox)*
+    - binaries and libs of cilium-envoy *(builded from src)*
```
Comment (apolovov) [inline modules/021-cni-cilium/images/README.md:50]: add a note about build cache in fox.

File: modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+# Original IMAGE_TAG of cilium/iptables is 67f517af50e18f64cd12625021f1c39246bb4f92
+# Corresponding commit is 3e6b0f588d788ac0e69f4d126cfb3c55b1513948 (https://github.com/cilium/image-tools/tree/3e6b0f588d788ac0e69f4d126cfb3c55b1513948)
+# #####################################################################
+---
+image: {{ $.ModuleName }}/base-cilium-dev
+from: {{ $.Images.BASE_UBUNTU }}
+fromCacheVersion: {{ $fromCacheVersion }}
+mount:
+- fromPath: ~/go-pkg-cache
+  to: /go/pkg
```
Comment (apolovov) [inline modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml:32]: remove pls :(

File: modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml [hunk 6]
```diff
@@ -0,0 +1,112 @@
+{{- $goVersion := "1.21.5" }}
+{{- $protocVersion := "22.3" }}
+{{- $bazelVersions := "3.7.0 3.7.1 3.7.2 6.1.0" }}
```
Comment (apolovov) [inline modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml:3]:
```suggestion
{{- $bazelVersions := list "3.7.0" "3.7.1" "3.7.2" "6.1.0" }}
```

File: modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+  - curl --fail --show-error --silent --location "https://github.com/protocolbuffers/protobuf/releases/download/v{{ $protocVersion }}/protoc-{{ $protocVersion }}-linux-x86_64.zip" --output /tmp/protoc.zip
+  - unzip /tmp/protoc.zip -x readme.txt -d /usr/local && rm /tmp/protoc.zip
+  - chmod o+rx /usr/local/bin/protoc && chmod o+rX -R /usr/local/include/google/
+  # 8ba23be9613c672d40ae261d2a1335d639bdd59b == tag: cmd/protoc-gen-go-grpc/v1.3.0
+  - go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@8ba23be9613c672d40ae261d2a1335d639bdd59b
+  - go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.30.0
+  - go install github.com/mitchellh/protoc-gen-go-json@v1.1.0
+  - go install github.com/pseudomuto/protoc-gen-doc/cmd/protoc-gen-doc@v1.5.1
+  # Install multiversion bazel
+  - bazel_versions=(3.7.0 3.7.1 3.7.2 6.1.0)
```
Comment (apolovov) [inline modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml:93]:
```suggestion
```

File: modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+  # install bazel wrapper script in the path, it automatically recognises `.bazelversion` and `USE_BAZEL_VERSIONS`, if neither are set it picks latest
+  - curl --fail --show-error --silent --location https://raw.githubusercontent.com/bazelbuild/bazel/${bazel_versions[-1]}/scripts/packages/bazel.sh --output /usr/local/bin/bazel
+  - chmod +x /usr/local/bin/bazel
+  - |
+    for bazel_version in "${bazel_versions[@]}" ; do
```
Comment (apolovov) [inline modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml:98]:
```suggestion
    for bazel_version in ${bazel_versions[@]} ; do
```

File: modules/021-cni-cilium/images/bin-cilium/werf.inc.yaml [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  - cd /go/src/github.com/cilium/cilium/images/runtime
+  - mkdir -p orig && cp -a iptables-wrapper-installer.sh orig/
+  - ./iptables-wrapper-installer.sh --no-sanity-check
+  setup:
+  - export GOROOT=/usr/local/go GOPATH=/go
+  - export PATH=${GOROOT}/bin:${GOPATH}/bin:${PATH}
+  - export GO_VERSION=${GOLANG_VERSION} GOPROXY={{ $.GOPROXY }}
+  - export GOOS=linux GOARCH=amd64 CGO_ENABLED=0
+  - cd /go/src/github.com/cilium/cilium
+  - find /patches -name '*.patch' -exec git apply {} \;
```
Comment (apolovov) [inline modules/021-cni-cilium/images/bin-cilium/werf.inc.yaml:62]:
```suggestion
  - find /patches -name '*.patch' | xargs git apply --verbose
```

File: modules/021-cni-cilium/images/agent/werf.inc.yaml [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+  - ln -f -s /run "/relocate/var/run"
+  - ln -f -s /run/lock "/relocate/var/lock"
+  #
+  - mkdir -p /relocate/home/cilium
+  - mkdir -p /relocate/etc
+  - cp -a /from_common_distroless/group /relocate/etc/group
+  - echo "cilium:x:1000:" >> /relocate/etc/group
+---
+# #####################################################################
+# New Main Ageng Image (Distroless)
```
Comment (apolovov) [inline modules/021-cni-cilium/images/agent/werf.inc.yaml:184]:
```suggestion
# New Main Agent Image (Distroless)
```

File: modules/021-cni-cilium/images/base-cilium-dev/binary_replace.sh (hunk 2 above)
Comment (AndreyPavlovFlant) [reply modules/021-cni-cilium/images/base-cilium-dev/binary_replace.sh]: `binary_replace.sh` deleted from this image.

File: modules/021-cni-cilium/images/README.md (hunk 4 above)
Comment (AndreyPavlovFlant) [reply modules/021-cni-cilium/images/README.md:52]: fixed

File: modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml (hunk 7 above)
Comment (AndreyPavlovFlant) [reply modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml]: fixed

File: modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml (hunk 8 above)
Comment (AndreyPavlovFlant) [reply modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml]: fixed

File: modules/021-cni-cilium/images/bin-cilium/werf.inc.yaml (hunk 9 above)
Comment (AndreyPavlovFlant) [reply modules/021-cni-cilium/images/bin-cilium/werf.inc.yaml]: fixed

File: modules/021-cni-cilium/images/agent/werf.inc.yaml (hunk 10 above)
Comment (AndreyPavlovFlant) [reply modules/021-cni-cilium/images/agent/werf.inc.yaml]: fixed

File: modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml (hunk 5 above)
Comment (AndreyPavlovFlant) [reply modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml:32]: It has been decided to leave this for now.

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/static'...

**Review by apolovov — APPROVED** [review]

**Review by RomanenkoDenys — APPROVED** [review]

**Pull request merged**